### PR TITLE
StopWordsRemovingEstimator export to Onnx

### PR DESCRIPTION
**Pull request by Lynx1820**
- Exporting StopWordsRemovingEstimator/CustomStopWordsRemovingEstimator to Onnx. 
- A test which currently fails is the edge case where all the words in the text get removed. While Ml.net produces an empty array, Onnx produces an array of length 1 with the empty string. Suggestions for how handle this case? 

Note: The decision was address the issue mentioned above in another PR.

### CODE REVIEW

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 1]
```diff
@@ -438,6 +441,45 @@ protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, b
...
+                    var dstVariableName = ctx.AddIntermediateVariable(_types[i], _parent.ColumnPairs[i].outputColumnName);
+                    SaveAsOnnxCore(ctx, i, srcVariableName, dstVariableName);
+                }
+            }
+
+            private void SaveAsOnnxCore(OnnxContext ctx, int iinfo, string srcVariableName, string dstVariableName)
+            {
+                var opType = "Squeeze";
+                var squeezeOutput = ctx.AddIntermediateVariable(null, "SqueezeOutput", true);
+                var node = ctx.CreateNode(opType, srcVariableName, squeezeOutput, ctx.GetNodeName(opType), "");
```
Comment (harishsk) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:463]: Is it possible to avoid skipping the shape and type addition?

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 2]
```diff
@@ -438,6 +441,45 @@ protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, b
...
+                node.AddAttribute("axes", new long[] { 0 });
```
Comment (harishsk) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:464]: Not in this PR, but in a different PR it maybe worth considering changing the default domain for CreateNode to be "ai.onnx" and not "ai.onnx.ml". The latter has very few ops and we mostly use operators from "ai.onnx" and it makes sense to retain that as the default.

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 3]
```diff
@@ -438,6 +441,45 @@ protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, b
...
+
+                opType = "StringNormalizer";
+                var stringNormalizerOutput = ctx.AddIntermediateVariable(null, "StringNormalizerOutput", true);
+                node = ctx.CreateNode(opType, squeezeOutput, stringNormalizerOutput, ctx.GetNodeName(opType), "");
```
Comment (harishsk) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:468]: Please try to add explicit types in all calls to AddIntermediateVariable.

**Comment by harishsk** [comment]
Can you check whether the number of elements returned is the same?

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs (hunk 1 above)
Comment (Lynx1820) [reply src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:463]: The reason I skipped the shape is because the shape of the tokenized word vector is not known prior to inference time. The number of words tokenized may be any number.

Comment (Lynx1820) [reply src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:463]: Other text transformers that also infer shape: 
- https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Transforms/Text/WordTokenizing.cs
- https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Transforms/Text/TokenizingByCharacters.cs

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs (hunk 3 above)
Comment (Lynx1820) [reply src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:468]: I know that types don't make it to the model if skip (mapping dataviews) is set to true. Should we still add types in that case for readability?

**Comment by Lynx1820** [comment]
> Can you check whether the number of elements returned is the same?

The length of the vectors is checked as part of the test. If you're referring to the edgecase where all the word tokens get removed, that test is not added here and it does failed because the lengths are not the same.

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs (hunk 3 above)
Comment (harishsk) [reply src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:468]: Yes, as far as possible please add types and shapes. 

---
In reply to: [449282071](https://github.com/dotnet/machinelearning/pull/5279#discussion_r449282071) [](ancestors = 449282071)

**Comment by harishsk** [comment]
Can you reshape the returned vector in SaveAsOnnx to be compatible with ML.NET?

---
In reply to: [653235748](https://github.com/dotnet/machinelearning/pull/5279#issuecomment-653235748) [](ancestors = 653235748)

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs (hunk 3 above)
Comment (harishsk) [reply src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:468]: If you add a type, but set skip=true, there is no impact. Do you get an error if you don't set skip=true?

---
In reply to: [449288767](https://github.com/dotnet/machinelearning/pull/5279#discussion_r449288767) [](ancestors = 449288767,449282071)

Comment (harishsk) [reply src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:468]: According to the documentation of [StringNormalizer](https://github.com/onnx/onnx/blob/master/docs/Operators.md#StringNormalizer) the output data type is a tensor of strings. So shouldn't the data type be `new VectorDataViewType(TextDataViewType)`?  (You have mentioned that you don't know the size ahead of time. Maybe you can leave it as `size=0`)

---
In reply to: [453111374](https://github.com/dotnet/machinelearning/pull/5279#discussion_r453111374) [](ancestors = 453111374,449288767,449282071)

**Comment by Lynx1820** [comment]
> Can you check whether the number of elements returned is the same?

I think this already done as part of the onnx testing framework: 
https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.TestFramework/BaseTestBaseline.cs#L751

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs (hunk 3 above)
Comment (Lynx1820) [reply src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:468]: I see. This is what I was talking about this morning. I see now that you mean adding just adding the type info while not skipping the mapping. I think this may work if the dimension remains the same, but since the dimension in this case switches from 2-d to 1-d back to 2-d, OnnxRuntime complains because it's expecting the data to stay 2-d. Even when replacing `_type[iinfo]` with `new VectorDataViewType(TextDataViewType)`, the issue remains.

Comment (harishsk) [reply src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:468]: Ok. Let us revisit this after this release.

---
In reply to: [453120111](https://github.com/dotnet/machinelearning/pull/5279#discussion_r453120111) [](ancestors = 453120111)
